src/pyqasm/algorithms/solovay_kitaev/solovay_kitaev.py

```diff
@@ -0,0 +1,147 @@
+        
+        return SU2Matrix(matrix, name)
+
+    def distance(self, other: 'SU2Matrix') -> float:
+        """Calculates the operator norm distance between two matrices."""
+        diff = self.matrix - other.matrix
+        return np.linalg.norm(diff)
+    
+    def _get_dagger_gate_name(self,name: str):
+        if name in SELF_INVERTING_ONE_QUBIT_OP_SET:
+            return name
+        else:
+            return ST_GATE_INV_MAP[name]
+        
+    def __str__(self):
+        return f"name: {self.name}, matrix: {self.matrix}"
+
+def group_commutator(a: SU2Matrix, b: SU2Matrix) -> SU2Matrix:
+    """Compute the group commutator [a,b] = aba^{-1}b^{-1}."""
+    return a * b * a.dagger() * b.dagger()        
+
+def find_basic_approximation(U: SU2Matrix, target_basis_set, accuracy=0.001, max_tree_depth=10) -> SU2Matrix:
+    gates = basic_approximation(U, target_basis_set, accuracy, max_tree_depth)
+    return SU2Matrix(gates["matrix"], gates["name"])
+
+def decompose_group_element(target: SU2Matrix, target_gate_set, basic_gates: List[SU2Matrix], depth: int) -> Tuple[List[SU2Matrix], float]:
+    
+    if depth == 0:
+        best_approx = find_basic_approximation(target.matrix, target_gate_set)
+        return best_approx, target.distance(best_approx)
+    
+    # Recursive approximation
+    prev_sequence, prev_error = decompose_group_element(target, target_gate_set, basic_gates, depth - 1)
+    
+    # If previous approximation is good enough, return it
+    # ERROR IS HARD CODED RIGHT NOW -> CHANGE THIS TO FIT USER-INPUT
+    if prev_error < 1e-6:
+        return prev_sequence, prev_error
+    
+    error = target * prev_sequence.dagger()
+    
+    # Find Va and Vb such that their group commutator approximates the error
+    best_v = None
+    best_w = None
+    best_error = float('inf')
+
+    for v in basic_gates:
+        for w in basic_gates:
+            comm = group_commutator(v, w)
+            curr_error = error.distance(comm)
+            if curr_error < best_error:
+                best_error = curr_error
+                best_v = v
+                best_w = w
+                
+    result = prev_sequence
+    
+    # Add correction terms
+    if best_v is not None and best_w is not None:
+        v_sequence, error  = decompose_group_element(best_v, target_gate_set, basic_gates, depth - 1)
+        w_sequence, error = decompose_group_element(best_w, target_gate_set, basic_gates, depth - 1)
+
+        result = group_commutator(v_sequence, w_sequence) * prev_sequence
+        
+    final_error = target.distance(result)
+    
+    return result, final_error
+
+def solovay_kitaev(target: np.ndarray, target_basis_set, depth: int = 3) -> List[np.ndarray]:
+    """
+    Main function to run the Solovay-Kitaev algorithm.
+    
+    Args:
+        target: Target unitary matrix as numpy array
+        target_basis_set: The target basis set to rebase the module to.
+        depth: Recursion depth
```

Comment:
Can we try and find some sort of relation between the recursion depth and the depth of the final sequence of gates? It'll be helpful for the user to set a particular depth limit for a gate decomposition